### DISCUSSION TITLE
Enums types should use snake case instead of simply lower case

### DISCUSSION THREAD

**Discussion by ghost** [sqlalchemy/sqlalchemy#10709] [Ideas]
If I have a Python enum called `MyCoolEnum`, when I use SQLAlchemy's `Enum(MyCoolEnum)`, this would get converted to an SQL type called "mycoolenum". I think it would be more readable if this were converted to `my_cool_enum`. 

Alternatively, is it possible to include `__typename__` (akin to `__tablename__` in tables) for enums?

**Comment by CaselIT** [comment]
Hi,

Unfortunately at the moment the name is hard coded to lower when not provided https://github.com/sqlalchemy/sqlalchemy/blob/66be1482db06adb908432b2e3b41d9393d1319f7/lib/sqlalchemy/sql/sqltypes.py#L1480-L1481

changing the default naming schema is likely a fairly disruptive change for all existing users so this would be opt in. It seems that `__typename__` could be an option since it's ignored by enum.

You can of course provide the name in the enum constructor

@zzzeek do you think this is something that may be useful?

**Reply by zzzeek** [reply]
I would not do anything with the user-defined enum class, magic naming conventions or anything like that.   rules for making custom naming would be within the Enum or events.

you can do this right now, not super cleanly but more or less, using schema events, since the name here is part of `Enum`'s participation in schema DDL:

```py
import enum

import sqlalchemy
from sqlalchemy import event
from sqlalchemy.dialects import postgresql
from sqlalchemy.dialects.postgresql import CreateEnumType
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column


# currently this recipe, to work fully with the declarative type map
# and all that, will only work with the base Enum class, not a custom
# subclass
#
# the point is I would want a function defined like this associated with
# the type_annotation_map, nothing magical on the enum.Enum class itself
@event.listens_for(sqlalchemy.Enum, "before_parent_attach")
def receive_before_parent_attach(target, parent):
    if target.name == "ASSIGN":
        target.name = target.enum_class.__name__


class Base(DeclarativeBase):
    type_annotation_map = {
        enum.Enum: sqlalchemy.Enum(enum.Enum, name="ASSIGN")
    }

# how it might work - assign the event to an instance

# base_enum = sqlalchemy.Enum(enum.Enum)

# @event.listens_for(base_enum, "before_parent_attach")
# def receive_before_parent_attach(target, parent):
#    target.name = target.enum_class.__name__

# the event would get used in this use case
# class Base(DeclarativeBase):
#     type_annotation_map = {
#         enum.Enum: base_enum
#     }


class Status(enum.Enum):
    PENDING = "pending"
    RECEIVED = "received"
    COMPLETED = "completed"


class SomeClass(Base):
    __tablename__ = "some_table"

    id: Mapped[int] = mapped_column(primary_key=True)
    status: Mapped[Status]


print(
    CreateEnumType(SomeClass.__table__.c.status.type).compile(
        dialect=postgresql.dialect()
    )
)

```
